src/CRFTDRegistry.sol

```diff
@@ -27,16 +27,27 @@ error ImplementationNotApproved();
 /// @notice CRFTD proxy registry
 contract CRFTDRegistry is Owned(msg.sender) {
     event Registered(address indexed user, bytes32 id, uint256 fee);
+    event TokenSetRegistered(address indexed user, bytes32 id, uint256 fee);
     event ProxyDeployed(address indexed owner, address indexed proxy);
 
+    uint256 public tokenRegisterFee = 0.0001 ether;
+
     mapping(address => bool) public approvedImplementation;
 
-    /* ------------- external ------------- */
+    mapping(bytes32 => bool) public paidStatus;
 
+    /* ------------- external ------------- */
     function register(bytes32 id) external payable {
         emit Registered(msg.sender, id, msg.value);
     }
 
+    function registerTokenSet(bytes32 tokenSetId, uint256 collectionSize) external payable {
+        uint256 fee = tokenRegisterFee * collectionSize;
+        if (msg.value != fee) revert IncorrectValue();
+        paidStatus[tokenSetId] = true;
+        emit TokenSetRegistered(msg.sender, tokenSetId, msg.value);
+    }
```

Comment:
here `tokenSetId = keccak256(abi.encode(msg.sender, collectionSize, nonce))` so it will be fine.

Reply:
wait, but this would also have to be enforced here in the registry then right? Atm it's possible to set `paidStatus[tokenSetId] = true;` on any arbitrary, user-controlled `tokenSetId` for a fee of `0`.

Reply:
I understood your point updated condition.

Reply:
Here we charging `eth` for the art generation and uploading stuff on IPFS. After emited event  we add extra validation step on front-end side. And here `nonce` is not just incrementing but it sometype of unique value for identifier.

Reply:
What is the "extra validation step on the frontend"? Does it check for `registry.paidStatus(tokenSetId)` whether it's `true` or does it sync the emitted events and check the `msg.value`?

Reply:
Yup. we can check emitted event or paidStatus before generating art.

Reply:
So, but why not perform the check correctly and not have it susceptible to attacks if using `paidStatus`?
I.e. why don't you write `tokenSetId = keccak256(abi.encode(msg.sender, collectionSize, nonce))` so that the control is out of the user's hands and they can't pre-image attack a desired `tokenSetId`? Also where exactly is the `nonce` stored?

Reply:
So are you suggesting like this
```
   uint256 nonce;

   function registerTokenSet(uint256 collectionSize) external payable {
        uint256 fee = tokenRegisterFee * collectionSize;
        if (msg.value != fee) revert IncorrectValue();
        bytes32 tokenSetId = keccak256(abi.encode(msg.sender, collectionSize, nonce++);
        paidStatus[tokenSetId] = true;
        emit TokenSetRegistered(msg.sender, tokenSetId, msg.value);
    }
```

Reply:
yes, but the `nonce` should be a mapping per user ideally. Though it might also make sense to have the backend first create a nonce (or id or random hash) once the user clicks on a new collection and let the user provide the nonce to the function. The function should probably also revert if the `paidStatus` is already `true`.